rancher/ui
# AKS updates: maxSurge, node pool labels, node pool taints, system pools

Addresses: https://github.com/rancher/dashboard/issues/8073
Addresses: https://github.com/rancher/dashboard/issues/8105

For AKS provisioned clusters - this PR:

- Adds support for maxSurge on Node Pools
- Adds support for Labels on Node Pools
- Adds support for Taints on Node Pools
- Adds support for adding system Node Pools
- Removes 'Advanced' section in Networking, so that 'Enable Private Cluster' is more discoverable
- Fixes a bug where not entering a node pool name left the UI stuck in "Saving..."

Notes:
- Taints are persisted as an array of strings in the format: `key=value:effect` - we parse this and format them this way on change
- Lablels and Taints sections will auto-expand on load if labels or taints are set

**To Test**

- Add a cloud credential for Azure
- Go to `Cluster Management`-> `Clusters` -> `Create`. Click `Azure AKS` under "Create a cluster in a hosted Kubernetes provider"
- Enter a cluster name, select your Azure cloud credential and click `Next: Configure Cluster`.
- Choose a Location and click `Configure Cluster`
- Enter a name for the `Cluster Resource Group` and `DNS Prefix`
- Scroll down to `Node Pools`
- You should see `Max Surge`, `Labels` and `Taints` are now available, e.g.

<img width="1271" alt="image" src="https://user-images.githubusercontent.com/1955897/231982164-0e969848-262f-4a2b-970d-2ea99a189ebd.png">

- Enter "20%" for Max Surge
- Add two labels: key: `label1`, value `value1` and key: `label2`, value `value2`
- Add a taint: key: `taint1`, value `taintvalue1`, effect: `PreferNoSchedle`

Should look like this:

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/1955897/231982779-d2a7c231-e153-4dae-bfa9-d19330f4eced.png">

- Click 'Add Node Pool`
- Enter the name `systempool`, set the Mode to `System` and select just Availability Zone 1.

- Click 'Add Node Pool`
- Enter the name `userpool`, set the Mode to `User` and select just Availability Zone 1.
- Enter "15%" for Max Surge
- Add a label: key: `userlabel1`, value `uservalue1`
- Add a taint: key: `usertaint1`, value `usertaintvalue1`, effect: `NoExecute`

- Create the cluster by clicking `Create`
- Wait for the cluster to  finish provisioning

- Open the Azure portal UI go to "Kubernetes Services" and locate the cluster you created above.
- Validate that the node pool settings made above are reflected in the Azure UI
- example of how node pools list should look:
![image](https://user-images.githubusercontent.com/1955897/231986864-16bfa196-2a41-45ae-b8d3-fd19dd3bb991.png)

Go through each Node Pool checking its Mode, Max Surge, Labels and Taints.

agentpool
<img width="1307" alt="image" src="https://user-images.githubusercontent.com/1955897/231987039-1e74e746-4b7b-44e6-8ab1-ca244ae48065.png">

systempool
<img width="1307" alt="image" src="https://user-images.githubusercontent.com/1955897/231987212-14abce6c-3656-4448-b15f-d05bb2b790ca.png">

userpool
<img width="1307" alt="image" src="https://user-images.githubusercontent.com/1955897/231987279-261f26be-2add-408b-badc-1133057c51c1.png">

- Back in Rancher in the Clusters list, `Edit Config` for the cluster created above
- Click `Next: Configure Cluster`
- Validate that the node pool data is shown correctly
- Validate that for the node pools with taints and labels, these sections are shown expanded
- Remove a label and taint from one of the pool. Change the Max Surge.
- Click Save
- Validate that the changes made are reflected when going back into edit in Rancher
- Validate that the changes made are reflected in the Azure Portal UI
- Delete the cluster
